[Dependencies.swift] Generate `DependenciesGraph` for `Carthage` dependencies.

laxmorek opened the pull request:
Part of #1674
Resolves https://github.com/tuist/tuist/issues/3001

### Short description 📝

This PR adds logic that generates the `graph.json` file for the `Carthage` dependencies. The `graph.json` is generated when user calls a `tuist dependencies fetch/update` command. 

### Example 🧐

An example of usage can be seen in the `app_with_framework_and_tests_and_dependencies` fixture. 

For the `Dependencies.swift` file:

```swift
let dependencies = Dependencies(
    carthage: .carthage(
        [
            .github(path: "ReactiveX/RxSwift", requirement: .exact("5.1.2")),
        ]
    )
    platforms: [.iOS]
)
```

The `graph.json` is generated as: 

```json
{
  "nodes" : {
    "RxTest" : {
      "kind" : "xcframework",
      "path" : "\/Tuist\/Dependencies\/Carthage\/RxTest.xcframework",
      "architectures" : [
        "armv7",
        "i386",
        "arm64",
        "x86_64"
      ]
    },
    "RxBlocking" : {
      "kind" : "xcframework",
      "path" : "\/Tuist\/Dependencies\/Carthage\/RxBlocking.xcframework",
      "architectures" : [
        "x86_64",
        "armv7k",
        "armv7",
        "arm64",
        "arm64_32",
        "i386"
      ]
    },
    "RxRelay" : {
      "kind" : "xcframework",
      "path" : "\/Tuist\/Dependencies\/Carthage\/RxRelay.xcframework",
      "architectures" : [
        "arm64",
        "x86_64",
        "armv7k",
        "i386",
        "armv7",
        "arm64_32"
      ]
    },
    "RxCocoa" : {
      "kind" : "xcframework",
      "path" : "\/Tuist\/Dependencies\/Carthage\/RxCocoa.xcframework",
      "architectures" : [
        "i386",
        "arm64",
        "arm64_32",
        "armv7k",
        "armv7",
        "x86_64"
      ]
    },
    "RxSwift" : {
      "kind" : "xcframework",
      "path" : "\/Tuist\/Dependencies\/Carthage\/RxSwift.xcframework",
      "architectures" : [
        "i386",
        "x86_64",
        "arm64",
        "armv7",
        "armv7k",
        "arm64_32"
      ]
    }
  }
}
```

### Implementation 👨‍💻

- `tuist dependencies fetch/update` command is called
- `tuist` installs `Carthage` dependencies 
- **[NEW]** - once dependencies is installed in a temporary directory `tuist` iterates over `.version` files and decode them as `CarthageVersionFile` models
- **[NEW]** - using information saved in `CarthageVersionFile` `tuist` generates `DependenciesGraph`
- **[NEW]** - `tuist` encodes and saves `DependenciesGraph` as a `graph.json` file
- `tuist` saves dependencies the `graph.json` file in `Tuist/Dependencies/` directory

You can read more about `.version` files [here](https://github.com/Carthage/Carthage/blob/master/Documentation/VersionFile.md)

### Questions ❓

- ~~In this PR I assumed that `Carthage` installs dependencies as `xcframework`s. We have to decide if we want to deprecate `.framework`s for `Carthage` installations or handle them. More [here](https://github.com/tuist/tuist/issues/2981#issuecomment-850526334).~~
- ~~I created `TuistCore.DependenciesGraphController` class that loads and saves `graph.json`. Im wondering if `TuistCore` is a good place for this class and Im still looking for better name. `DependenciesGraphCacher`, `DependenciesGraphLoader`?~~

luispadron commented on Sources/TuistCore/DependenciesGraph/DependenciesGraphController.swift:
```diff
@@ -0,0 +1,58 @@
+    /// - Parameters:
+    ///   - dependenciesGraph: A model that will be saved.
+    ///   - path: Directory whose project's dependencies graph will be saved.
+    func save(_ dependenciesGraph: DependenciesGraph, at path: AbsolutePath) throws
```
```suggestion
    func save(_ dependenciesGraph: DependenciesGraph, to path: AbsolutePath) throws
```

luispadron commented on Sources/TuistCore/DependenciesGraph/DependenciesGraphController.swift:
```diff
@@ -0,0 +1,58 @@
+    // MARK: - Helpers
+
+    private func graphPath(at path: AbsolutePath) -> AbsolutePath {
+        path
+            .appending(component: Constants.tuistDirectoryName)
+            .appending(component: Constants.DependenciesDirectory.name)
+            .appending(component: Constants.DependenciesDirectory.graphName)
```
nit: you can use `appending(components: ..., ..., ...)` instead

luispadron commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -52,27 +52,19 @@ enum CarthageInteractorError: FatalError, Equatable {
 
 public protocol CarthageInteracting {
-    /// Fetches `Carthage` dependencies.
+    /// Installes `Carthage` dependencies
```
```suggestion
    /// Installs `Carthage` dependencies
```

luispadron commented on Sources/TuistDependencies/Carthage/Models/CarthageVersionFile.swift:
```diff
@@ -0,0 +1,50 @@
+        let identifier: String
+
+        /// Returns architectures the product is built for.
+        var architectures: [BinaryArchitecture] {
+            // example identifier: `ios-arm64_i386_x86_64-simulator`
+            identifier
+                .components(separatedBy: "-")[1]
+                .replacingOccurrences(of: "x86_64", with: "x8664")
+                .replacingOccurrences(of: "arm64_32", with: "arm6432")
+                .components(separatedBy: ["_"])
+                .map { $0 == "x8664" ? "x86_64" : $0 }
+                .map { $0 == "arm6432" ? "arm64_32" : $0 }
+                .compactMap { BinaryArchitecture(rawValue: $0) }
+        }
+    }
```
This seems a bit fragile, we should make sure its well tested. I'd also guard the `.components(separatedBy: "-")[1]` incase the list is empty

laxmorek replied:
> I'd also guard the .components(separatedBy: "-")[1] incase the list is empty

Good idea

> we should make sure its well tested

I have already added a few tests [here](https://github.com/tuist/tuist/blob/66814e20a6f2da122b5d84086d5e53cf07fac6f2/Tests/TuistDependenciesTests/Carthage/Models/CarthageVersionFileTests.swift)

luispadron reviewed (commented):
This is looking great!

fortmarek commented on Sources/TuistCore/DependenciesGraph/DependenciesGraphController.swift:
```diff
@@ -0,0 +1,58 @@
+    /// Saves the `DependenciesGraph` as `graph.json`.
+    /// - Parameters:
+    ///   - dependenciesGraph: A model that will be saved.
+    ///   - path: Directory whose project's dependencies graph will be saved.
```
`whose` x `where`?

fortmarek commented on Sources/TuistCore/DependenciesGraph/DependenciesGraphController.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public func load(at path: AbsolutePath) throws -> DependenciesGraph {
+        let graphPath = graphPath(at: path)
```
I believe this won't build on Xcode 12.4

laxmorek replied:
yes it won't 😅 do we still support Xcode 12.4? what is the lowest version that we should support? 🤔

laxmorek replied:
ok, I can see that CI is failed because of that, I will fix it

fortmarek replied:
we should be conservative with dropping Xcode versions as not all teams move that quickly (keeping the last few minor versions usually). Plus in this case, Xcode 12.5 is still not supported by Github Actions.

fortmarek commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -52,27 +52,19 @@ enum CarthageInteractorError: FatalError, Equatable {
-    func fetch(
+    ///   - dependencies:  List of dependencies to install using `Carthage`.
+    ///   - platforms: List of platforms for which you want to install dependencies.
+    ///   - shouldUpdate: Indicates whether dependencies should be updated or fetched basing on the `Tuist/Lockfiles/Cartfile.resolved` lockfile.
```
```suggestion
    ///   - shouldUpdate: Indicates whether dependencies should be updated or fetched based on the `Tuist/Lockfiles/Cartfile.resolved` lockfile.
```

fortmarek commented on Sources/TuistDependencies/Carthage/Models/CarthageVersionFile.swift:
```diff
@@ -0,0 +1,50 @@
+import Foundation
+import TuistGraph
+
+/// A model that represents the `Carhtage` version file.
```
```suggestion
/// A model that represents the `Carthage` version file.
```

fortmarek commented on Sources/TuistDependencies/Carthage/Models/CarthageVersionFile.swift:
```diff
@@ -0,0 +1,50 @@
+                .replacingOccurrences(of: "arm64_32", with: "arm6432")
+                .components(separatedBy: ["_"])
+                .map { $0 == "x8664" ? "x86_64" : $0 }
+                .map { $0 == "arm6432" ? "arm64_32" : $0 }
```
this seems redundant with the `replacingOccurrences` functions above?

laxmorek replied:
I have to do that to escape `arm64_32` and `x86_64` to split the string by `_` 

once string is split I have to restore original format to meet [TuistGraph.BinaryArchitecture](https://github.com/tuist/tuist/blob/ba28383dc89aab1d411002c0303fe794e24db767/Sources/TuistGraph/Models/BinaryArchitecture.swift#L3)'s raw values.

fortmarek replied:
Right, the separator of `_` used by carthage is not the best. Let's keep it as-is

fortmarek commented on Sources/TuistCore/DependenciesGraph/DependenciesGraphController.swift:
```diff
@@ -0,0 +1,58 @@
+import TuistSupport
+
+/// A protocol that defines an interface to save and load the `DependenciesGraph` using a `graph.json` file.
+public protocol DependenciesGraphControlling {
```
Should this be in `TuistCore`? Feels tied to dependencies, rather

fortmarek commented on Sources/TuistDependencies/Carthage/Utils/CarthageGraphGenerator.swift:
```diff
@@ -0,0 +1,51 @@
+    private let fileHandler: FileHandling
+
+    public init(
+        fileHandler: FileHandling = FileHandler.shared
```
nit: it is not necessary to inject `FileHandler.shared`. Actually, the whole point of `FileHandler.shared` instance is that it does not injected. Feel free to use it directly.

fortmarek commented on Sources/TuistDependencies/Carthage/Models/CarthageVersionFile.swift:
```diff
@@ -0,0 +1,50 @@
+        let identifier: String
+
+        /// Returns architectures the product is built for.
+        var architectures: [BinaryArchitecture] {
```
I was wondering what do we use this for? Can we make some validations based on this or?

laxmorek replied:
> Can we make some validations based on this or? 

Maybe... 🤔 

> I was wondering what do we use this for?

I assumed it will be mapped into something similar to `ValueGraphDependency.xcframework` when we start to marge graphs.

The `ValueGraphDependency.xcframework` contains the `XCFrameworkInfoPlist` and `XCFrameworkInfoPlist` contains [architectures](https://github.com/tuist/tuist/blob/ba28383dc89aab1d411002c0303fe794e24db767/Sources/TuistGraph/Models/XCFrameworkInfoPlist.swift#L30) property.  

So if we can collect this information from `Carthage` installation why do not store it. 🤷‍♂️

Anyway, if you have a feeling it will be not necessary or it could be added later then I can remove it.

fortmarek replied:
good point, we might need for the xcframework plist, let's keep it 👍

fortmarek commented on Sources/TuistDependencies/Carthage/Utils/CarthageGraphGenerator.swift:
```diff
@@ -0,0 +1,51 @@
+            .flatMap { $0.allProducts }
+
+        let nodes = Dictionary(grouping: products, by: { $0.name })
+            .reduce(into: [String: DependenciesGraphNode]()) { result, next in
```
nit: Rather than `next`, I'd rename this to something more explicit (I suppose `product`?)

fortmarek commented on Sources/TuistDependencies/Carthage/Utils/CarthageGraphGenerator.swift:
```diff
@@ -0,0 +1,51 @@
+                guard let frameworkName = next.value.first?.container else { return }
+
+                let path = AbsolutePath("/")
+                    .appending(component: Constants.tuistDirectoryName)
```
nit: I would use `.appending(components: ...)`

fortmarek commented on Sources/TuistDependencies/SwiftPackageManager/SwiftPackageManagerInteractor.swift:
```diff
@@ -33,25 +33,16 @@ enum SwiftPackageManagerInteractorError: FatalError, Equatable {
 
 public protocol SwiftPackageManagerInteracting {
-    /// Fetches `Swift Package Manager` dependencies.
+    /// Installes `Swift Package Manager` dependencies.
```
```suggestion
    /// Installs `Swift Package Manager` dependencies.
```

fortmarek commented on Sources/TuistGraph/DependenciesGraph/DependenciesGraph.swift:
```diff
@@ -0,0 +1,16 @@
+/// A directed acyclic graph (DAG) that Tuist uses to represent the third party dependency tree.
+public struct DependenciesGraph: Equatable, Codable {
+    /// A dictionary where the keys are the names of dependencies,
+    /// and the values are the dependencies itself.
```
```suggestion
    /// and the values are the dependencies themselves.
```

fortmarek commented on Sources/TuistGraph/DependenciesGraph/DependenciesGraph.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+import TSCBasic
+
+/// A directed acyclic graph (DAG) that Tuist uses to represent the third party dependency tree.
```
Right now this does not represent a DAG - but creating a DAG will be necessary for SPM, so feel free to keep it (we don't create edges between Carthage's dependencies as they are not defined)

fortmarek commented on Sources/TuistGraph/DependenciesGraph/DependenciesGraphNode.swift:
```diff
@@ -0,0 +1,43 @@
+import TSCBasic
+
+// A node of the third party dependency graph.
+public enum DependenciesGraphNode: Hashable, Equatable, Codable {
```
To be more in line with the graph used for the rest of the project, I'd maybe rename this to `ThirdPartyDependency`? Or something like that? `Node` suffix has been in use before the introduction of `ValueGraph`.

laxmorek replied:
It makes sense to me.

fortmarek commented on Sources/TuistGraph/DependenciesGraph/DependenciesGraph.swift:
```diff
@@ -0,0 +1,16 @@
+public struct DependenciesGraph: Equatable, Codable {
+    /// A dictionary where the keys are the names of dependencies,
+    /// and the values are the dependencies itself.
+    public var nodes: [String: DependenciesGraphNode]
```
I'd rename this to `thirdPartyDependencies`. `nodes` is IMHO not descriptive enough and for the actual DAG representation it will be necessary to create a new property that will be one-to-many (`[DependenciesGraphNode: [DependenciesGraphNode]]`)

laxmorek replied:
It makes sense to me.

I have expected some "graph" related feedback. Thank you :)

fortmarek commented:
> In this PR I assumed that Carthage installs dependencies as xcframeworks. We have to decide if we want to deprecate .frameworks for Carthage installations or handle them. More here.

The dependencies manifest is not yet stable and it's not marketed as such, so we can just remove the possibility of building regular frameworks. I think it's not necessary to support for the time being. IMHO, it's more important to have some support for Carthage and then moving on to SPM.

> I created TuistCore.DependenciesGraphController class that loads and saves graph.json. Im wondering if TuistCore is a good place for this class and Im still looking for better name. DependenciesGraphCacher, DependenciesGraphLoader?

Yeah, as mentioned in the review, I'd move it to `TuistDependencies` if possible. I think `DependenciesGraphController` is fine, can't think of a better name.

Anyways, this is looking good @laxmorek! I'm looking forward to another round of review.

laxmorek commented:
> The dependencies manifest is not yet stable and it's not marketed as such, so we can just remove the possibility of building regular frameworks.

Cool. I have the same feeling. Let's drop a support for regular frameworks. 

> (about `DependenciesGraphController`) Yeah, as mentioned in the review, I'd move it to `TuistDependencies` 

I added it to `TuistCore` because `DependenciesGraphController` can be used to load the `DependenciesGraph` from `graph.json`. So probably `TuistLoader` will use it during the `generate` command. I guess we don't want to add a dependency `TuistLoader ` -> `TuistDependencies`.

fortmarek commented:
> So probably TuistLoader will use it during the generate command. I guess we don't want to add a dependency TuistLoader -> TuistDependencies

While that might be true, you can load the dependencies graph using code `TuistDependencies` and then inject it to `TuistLoader`. IMHO, `TuistLoader` dioes not have to use it directly, thus `DependenciesGraphController` can safely live in `TuistDependencies` framework.

laxmorek commented:
> thus `DependenciesGraphController` can safely live in `TuistDependencies` framework

ok, lets try this approach

laxmorek commented:
@fortmarek @luispadron the PR is ready for another round of review.

danyf90 commented on CHANGELOG.md:
```diff
@@ -4,6 +4,18 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
+
+### Changed
+
+- **breaking** The `tuist dependencies` command requires the `Carthage` in version at least `0.37.0`. [#3043](https://github.com/tuist/tuist/pull/3043) by [@laxmorek](https://github.com/laxmorek)
```
```suggestion
- **breaking** The `tuist dependencies` command requires the `Carthage` version to be at least `0.37.0`. [#3043](https://github.com/tuist/tuist/pull/3043) by [@laxmorek](https://github.com/laxmorek)
```

danyf90 commented on Sources/ProjectDescription/Dependencies/CarthageDependencies.swift:
```diff
@@ -5,20 +5,13 @@ public struct CarthageDependencies: Codable, Equatable {
-    ///   - dependencies: List of depedencies that can be installed using Carthage.
-    ///   - options: List of options for Carthage installation.
+    /// - Parameter dependencies: List of depedencies that can be installed using Carthage.
     public static func carthage(
```
What about using the init instead? Given it will be passed to a `carthage` property, having to repeat the name is a bit strange, isn't it?
We would go from:
```
let dependencies = Dependencies(
    carthage: .carthage([
        …
    ]
)
)
```
to
```
let dependencies = Dependencies(
    carthage: .init(
        [
            …
        ]
    )
)
```

We could go even further, and make it `ExpressibleByArrayLiteral`, so it would be even clearer:
```
let dependencies = Dependencies(
    carthage: [
        …
    ],
    …
)
```

fortmarek replied:
> We could go even further, and make it ExpressibleByArrayLiteral, so it would be even clearer:

I believe this would be possible to do already. But yeah, I think the approach you have outlined might be more straightforward, so +1 from me.

laxmorek replied:
> What about using the init instead?

That was done on propose. The discussion about it can be found [here](https://github.com/tuist/tuist/pull/2532#issuecomment-781462676).

> We could go even further, and make it `ExpressibleByArrayLiteral`, so it would be even clearer:

👍

laxmorek replied:
Should we remove `static` builders and make `init` public?

danyf90 commented on Sources/TuistDependencies/Carthage/Utils/CarthageGraphGenerator.swift:
```diff
@@ -0,0 +1,47 @@
+            .map { try jsonDecoder.decode(CarthageVersionFile.self, from: $0) }
+            .flatMap { $0.allProducts }
+
+        let thirdPartyDependencies = Dictionary(grouping: products, by: { $0.name })
+            .reduce(into: [String: ThirdPartyDependency]()) { result, product in
+                guard let frameworkName = product.value.first?.container else { return }
+
+                let path = AbsolutePath("/")
+                    .appending(components: [
+                        Constants.tuistDirectoryName,
+                        Constants.DependenciesDirectory.name,
+                        Constants.DependenciesDirectory.carthageDirectoryName,
+                        frameworkName,
+                    ])
+
+                let architectures: Set<BinaryArchitecture> = Set(product.value.flatMap { $0.architectures })
+                result[product.key] = .xcframework(path: path, architectures: architectures)
+            }
```
A bit simpler 🤏
```suggestion
        let thirdPartyDependencies = Dictionary(grouping: products, by: \.name)
            .compactMapValues { product in
                guard let frameworkName = product.first?.container else { return nil }

                let path = AbsolutePath("/")
                    .appending(components: [
                        Constants.tuistDirectoryName,
                        Constants.DependenciesDirectory.name,
                        Constants.DependenciesDirectory.carthageDirectoryName,
                        frameworkName,
                    ])

                let architectures: Set<BinaryArchitecture> = Set(product.flatMap { $0.architectures })
                return .xcframework(path: path, architectures: architectures)
            }
```

laxmorek replied:
👍

danyf90 commented on Sources/TuistDependencies/DependenciesController.swift:
```diff
@@ -114,24 +119,34 @@ public final class DependenciesController: DependenciesControlling {
             throw DependenciesControllerError.noPlatforms
         }
 
+        var dependenciesGraph: DependenciesGraph?
```
Maybe it's easier to make it non-nillable, so that for each package manager we just need to merge the content with the output of the package manager, without any special handling of the first one.
Then, the last `if` just check whether it's empty instead of nil, which would also be a bit more robust, as now we might create an empty json in case dependenciesGraph is not nil but empty

laxmorek commented:
I updated the PR. The biggest changes since my last update is making `CarthageDependencies` and `SwiftPackageManagerDependencies` conformable to `ExpressibleByArrayLiteral`. @danyf90, thx for suggesting it. 

I updated `app_with_framework_and_tests_and_dependencies` and documentation to follow the newest approach.

danyf90 commented on Sources/TuistDependencies/DependenciesController.swift:
```diff
@@ -114,24 +119,30 @@ public final class DependenciesController: DependenciesControlling {
             try swiftPackageManagerInteractor.clean(dependenciesDirectory: dependenciesDirectory)
         }
+
+        try dependenciesGraphController.save(dependenciesGraph, to: path)
```
Shouldn't this be:
```suggestion
        if dependenciesGraph.thirdPartyDependencies.isEmpty {
            try dependenciesGraphController.save(dependenciesGraph, to: path)
        } else {
            try dependenciesGraphController.clean(at: path)
        }
```
?

If instead we want to write it anyway, then we can get rid of the clean method

luispadron reviewed (approved):
Loving the API update to allow conforming to array literal! This is looking great

laxmorek commented:
Hey @luispadron thank you for your review. I have added one more commit that addressing your [comment](https://github.com/tuist/tuist/pull/3043#discussion_r644782677).
